src/star-id.cpp

```diff
@@ -1,21 +1,338 @@
+                // std::cout << "Alert: matches size = 0, continuing" << std::endl;
+                continue;
+            }
+
+            for (std::vector<int> matchRow : matches) {
+                // Construct the pattern we found in the Pattern Catalog
+                std::vector<int> catStarIDs;
+                std::vector<Vec3> catStarVecs;
+                for (int star : matchRow) {
+                    CatalogStar catStar = catalog[tetraDatabase.GetTrueCatInd(star)];
+                    Vec3 catVec = catStar.spatial;
+                    catStarIDs.push_back(catStar.name);
+                    catStarVecs.push_back(catVec);
+                }
+
+                std::vector<float> catEdgeRatios = TetraConstructPattern(catStarVecs);
+
+                bool skipMatchRow = false;
+                for (int i = 0; i < (int)catEdgeRatios.size(); i++) {
+                    float val = catEdgeRatios[i] - pattEdgeRatios[i];
+                    val = std::abs(val);
+                    // Test that our pattern and PC pattern roughly match up
+                    // For now, just compare edge ratios
+                    if (val > pattMaxError) {
+                        skipMatchRow = true;
+                    }
+                }
+
+                // Very common to skip here, database Pattern is NOT a match to ours
+                if (skipMatchRow) {
+                    // std::cout << "Alert: not a match!" << std::endl;
+                    continue;
+                }
+
+                // Awesome! The PC pattern does match ours
+                // Reorder star vectors in both Patterns so we can match them up by ID
+
+                Vec3 pattCentroid(0, 0, 0);
+                for (Vec3 pattStarVec : pattStarVecs) {
+                    pattCentroid = pattCentroid + pattStarVec;
+                }
+                pattCentroid = pattCentroid * (1.0 / (int)pattStarVecs.size());
+
+                std::vector<float> pattRadii;
+                for (Vec3 pattStarVec : pattStarVecs) {
+                    pattRadii.push_back((pattStarVec - pattCentroid).Magnitude());
+                }
+
+                std::vector<int> sortedCentroidIndices =
+                    ArgsortVector<int, float>(chosenIndIndices, pattRadii);
+
+                std::vector<Vec3> pattSortedVecs =
+                    ArgsortVector<Vec3, float>(pattStarVecs, pattRadii);
+
+                Vec3 catCentroid(0, 0, 0);
+                for (Vec3 catStarVec : catStarVecs) {
+                    catCentroid = catCentroid + catStarVec;
+                }
+                catCentroid = catCentroid * (1.0 / (int)catStarVecs.size());
+
+                std::vector<float> catRadii;
+                for (Vec3 catStarVec : catStarVecs) {
+                    catRadii.push_back((catStarVec - catCentroid).Magnitude());
+                }
+
+                std::vector<int> catSortedStarIDs = ArgsortVector<int, float>(catStarIDs, catRadii);
```

Comment:
I think this whole bit at the end could be simplified a bit by just having two vectors of `std::pair` with radius as the first element of each pair, rather than four total vectors which you use with argsort.

By using pairs, you can then just use `std::sort`, because it automatically sorts pairs by the first element of the pair.

Reply:
I think here it might be easier to just go with Argsort, since even though I sort by radius each time the 2 vectors I sort on are used quite a bit beforehand, so it'd be annoying to put them in a pair just for sorting